### PR TITLE
AB#31: Elastic search

### PR DESCRIPTION
**Pull request by bytetum**
---

<a href="https://app.devin.ai/review/ez-map/ezmap-poc/pull/11" target="_blank">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open with Devin
  </picture>
</a>

### CODE REVIEW

File: EzMap.Api/Controllers/PoiCollectionController.cs [hunk 1]
```diff
@@ -16,13 +17,40 @@ public class PoiCollectionController : ControllerBase
...
+                return new StatusCodeResult(StatusCodes.Status500InternalServerError);
+            }
+            
+            var poiCollectionCreateIndex = new PoiCollectionCreateIndex(poiCollectionId, dto.Name, dto.Description);
+            var esResult = await elasticSearchService.AddOrUpdate(poiCollectionCreateIndex);
+            if (!esResult)
+            {
+                return Ok(new
+                {
+                    Message = "Your poi collection is created successfully, but indexing encountered an issue.",
```
Comment (popemkt) [inline EzMap.Api/Controllers/PoiCollectionController.cs]: You don't expose your implementation details to users/public apis in error messages.

If indexing failed is not an error, do you have any way to fix the the index, or will it just stay that way? In that case your system fell in to a state where it's not consistent, some records are indexed, and some aren't, for which you don't have any way to know or to recover from.

File: EzMap.Api/Controllers/PoiCollectionController.cs [hunk 2]
```diff
@@ -16,13 +17,40 @@ public class PoiCollectionController : ControllerBase
...
+        try
+        {
+            var poiCollectionId = uow.PoiCollectionRepository.AddPoiCollection(dto.WithUserId(identityService.GetUserId()));
+            var dbResult = await uow.SaveAsync();
+            if (dbResult <= 0)
+            {
+                return new StatusCodeResult(StatusCodes.Status500InternalServerError);
+            }
+            
+            var poiCollectionCreateIndex = new PoiCollectionCreateIndex(poiCollectionId, dto.Name, dto.Description);
```
Comment (popemkt) [inline EzMap.Api/Controllers/PoiCollectionController.cs]: Why is the model for indexing an entity called Index? Is it an index or is it a model to index the entity?

File: EzMap.Api/Services/ElasticSearchService.cs [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    private string _indexName { get; set; }
+    private readonly IElasticClient _client;
+
+    public ElasticSearchService(IElasticClient client)
+    {
+        _indexName = "thanh1";
+        _client = client;
+    }
+
+    public ElasticSearchService Index(string indexName)
```
Comment (popemkt) [inline EzMap.Api/Services/ElasticSearchService.cs]: Is it index, or is it CreateIndex? Or is it SetIndex?

File: EzMap.Api/Controllers/PoiCollectionController.cs (hunk 2 above)
Comment (bytetum) [reply EzMap.Api/Controllers/PoiCollectionController.cs]: The purpose of this is treat it as a model for indexing an entity

Comment (popemkt) [reply EzMap.Api/Controllers/PoiCollectionController.cs]: Then change the name pls

File: EzMap.Api/Services/ElasticSearchService.cs (hunk 3 above)
Comment (bytetum) [reply EzMap.Api/Services/ElasticSearchService.cs]: it is CreateIndex if it is not existed, refactoring this

File: EzMap.Api/Controllers/PoiCollectionController.cs [hunk 4]
```diff
@@ -16,13 +19,51 @@ public class PoiCollectionController : ControllerBase
...
+                {
+                    await transaction.RollbackAsync();
+                    return new StatusCodeResult(StatusCodes.Status500InternalServerError);
+                }
+                var poiCollectionCreateIndex = new PoiCollectionCreateIndexingModel(poiCollectionId, dto.Name, dto.Description);
+                var esResult = await elasticSearchService.AddOrUpdate(poiCollectionCreateIndex);
+                if (!esResult)
+                {
+                    await transaction.RollbackAsync();
+                    return Ok(new
```
Comment (popemkt) [inline EzMap.Api/Controllers/PoiCollectionController.cs]: Still Ok even if it fails?

File: EzMap.Api/Controllers/PoiCollectionController.cs [hunk 5]
```diff
@@ -16,13 +19,51 @@ public class PoiCollectionController : ControllerBase
...
+                    Id = poiCollectionId,
+                });
+            }
+            catch
+            {
+                await transaction.RollbackAsync();
+                throw;
+            }
+        }
+        catch (Exception ex)
```
Comment (popemkt) [inline EzMap.Api/Controllers/PoiCollectionController.cs]: Please create exception filter or middleware for this, this is a generic catch (with no errors specific to this particular endpoint, it should not be here)

File: EzMap.Api/Controllers/PoiCollectionController.cs [hunk 6]
```diff
@@ -46,37 +87,106 @@ public async Task<IActionResult> Search([FromServices] IUnitOfWork uow,
...
+                await transaction.CommitAsync();
+                return Ok("Your poi collection is updated successfully");
+            }
+            catch
+            {
+                await transaction.RollbackAsync();
+                throw;
+            }
+        }
+        catch (Exception ex)
```
Comment (popemkt) [inline EzMap.Api/Controllers/PoiCollectionController.cs]: same as above

File: EzMap.Api/Controllers/PoiCollectionController.cs [hunk 7]
```diff
@@ -46,37 +87,106 @@ public async Task<IActionResult> Search([FromServices] IUnitOfWork uow,
...
+            await using var transaction = await uow.BeginTransactionAsync();
+            try
+            {
+                var dbPoiCol = await uow.PoiCollectionRepository.GetPoiCollectionById(identityService.GetUserId(), dto.Id);
+
+                if (dbPoiCol is not null)
+                {
+                    uow.PoiCollectionRepository.UpdatePoiCollectionAsync(dbPoiCol, dto);
+                    var dbResult = await uow.SaveAsync();
+                    if (dbResult <= 0)
```
Comment (popemkt) [inline EzMap.Api/Controllers/PoiCollectionController.cs]: The only way for this to have result <= 0 is uow.PoiCollectionRepository.UpdatePoiCollectionAsync doesn't update anything, which is really wrong because in that case you need to fix UpdatePoiCollectionAsync, not throw, that just means there's a bug. Imagine you know a bug exists, and you just throws when it happens, instead of fixing it. This is the same. 

For cases where for example the db failed to connect, an exception is thrown and this line is never reached.

In short, do not check for every edge case (that won't happen) in a paranoid way, doing that will result in a check for every line you write.

File: EzMap.Api/Controllers/PoiCollectionController.cs [hunk 8]
```diff
@@ -46,37 +87,106 @@ public async Task<IActionResult> Search([FromServices] IUnitOfWork uow,
...
+                {
+                    return BadRequest("Please provide a valid id!");
+                }
 
-        return await uow.SaveAsync() > 0
-            ? Ok("Your poi collection is deleted successfully!")
-            : new StatusCodeResult(StatusCodes.Status500InternalServerError);
+                await uow.PoiCollectionRepository.DeletePoiCollectionAsync(id);
+                var dbResult = await uow.SaveAsync();
+                if(dbResult <= 0){
```
Comment (popemkt) [inline EzMap.Api/Controllers/PoiCollectionController.cs]: This doesn't work either. It's nice to check if a delete is successful, but when you can't find an item to delete, you return 404 not found, not 500 internal server error. Not every error is the app's fault, in this case it's the requester's fault for providing a non-existing id.

File: EzMap.Api/Controllers/PoiController.cs [hunk 9]
```diff
@@ -17,26 +19,64 @@ public class PoiController : ControllerBase
...
-            ? Ok("Your point of interest is created successfully!")
-            : new StatusCodeResult(StatusCodes.Status500InternalServerError);
+        try
+        {
+            await using var transaction = await uow.BeginTransactionAsync();
+            try 
+            {
+                var poiId = uow.PoiRepository.AddPoi(dto.WithUserId(identityService.GetUserId()));
+                var dbResult = await uow.SaveAsync();
+                if (dbResult <= 0)
```
Comment (popemkt) [inline EzMap.Api/Controllers/PoiController.cs]: This also doesn't work. Please review all these and fix if it makes sense or not to check for result and throw exception

File: EzMap.Api/Controllers/PoiController.cs [hunk 10]
```diff
@@ -17,26 +19,64 @@ public class PoiController : ControllerBase
...
 
     [Authorize]
     [HttpPut("{id:guid}")]
     public async Task<IActionResult> Update([FromBody] PoiUpdateDto dto, [FromServices] IUnitOfWork uow,
-        [FromServices] IIdentityService identityService)
+        [FromServices] IIdentityService identityService, [FromServices] IElasticSearchService elasticSearchService)
     {
         var dbPoi = await uow.PoiRepository.GetPoiById(identityService.GetUserId(), dto.Id);
 
         if (dbPoi is not null)
```
Comment (popemkt) [inline EzMap.Api/Controllers/PoiController.cs]: again, if can't find Poi that means it's 404.

File: EzMap.Api/Program.cs [hunk 11]
```diff
@@ -22,8 +42,9 @@
 builder.Services.AddScoped<IIdentityService, IdentityService>();
 builder.Services.AddFluentValidationAutoValidation();
 builder.Services.AddValidatorsFromAssemblyContaining<PoiCreateDtoValidator>();
-
-var appSettings = builder.Configuration.GetValue<string>("AppSecret");
+// builder.Environment.EnvironmentName = "Test";
```
Comment (popemkt) [inline EzMap.Api/Program.cs]: No extra comments

File: EzMap.Api/Services/ElasticSearchService.cs (hunk 3 above)
Comment (popemkt) [reply EzMap.Api/Services/ElasticSearchService.cs]: And why is it refactored and still here?

File: EzMap.Domain/Repositories/TagRepository.cs [hunk 12]
```diff
@@ -22,31 +23,39 @@ public class TagRepository : ITagRepository
...
+        
         return tag;
     }
 
-    public void AddTag(TagCreateDto dto)
+    public Guid AddTag(TagCreateDto dto)
     {
         Tag tag = new Tag(dto.Name, dto.Description, dto.UserId);
-        _dbContext.Tags.Add(tag);
+        _dbContext.Tags?.Add(tag);
```
Comment (popemkt) [inline EzMap.Domain/Repositories/TagRepository.cs]: Why do you even check for null here it doesn't even make sense.
Context collections are never null so never mark them as null, worst case they are EMPTY, not NULL. Ignore the IDE and compiler in this case.

File: EzMap.Domain/Repositories/PoiRepository.cs [hunk 13]
```diff
@@ -78,6 +92,13 @@ public async Task DeletePoiAsync(Guid id, CancellationToken token = default)
             pois = pois.Where(x => (x.Address.ToLower().Contains(keyword.ToLower()) || x.Name.ToLower().Contains(keyword.ToLower())) && x.UserId == userId ).ToList();
         }
 
+        if (!pois.Any())
+        {
+            _logger.LogInformation("This poi is not existing");
```
Comment (popemkt) [inline EzMap.Domain/Repositories/PoiRepository.cs]: These are multiple pois, not one. Please write the appropriate message

File: EzMap.Domain/Repositories/PoiRepository.cs [hunk 14]
```diff
@@ -78,6 +92,13 @@ public async Task DeletePoiAsync(Guid id, CancellationToken token = default)
             pois = pois.Where(x => (x.Address.ToLower().Contains(keyword.ToLower()) || x.Name.ToLower().Contains(keyword.ToLower())) && x.UserId == userId ).ToList();
```
Comment (popemkt) [inline EzMap.Domain/Repositories/PoiRepository.cs]: Why do you already get the the list from db then filter, you must filter at db otherwise you'll be getting all the pois in memory

File: EzMap.Domain/Repositories/PoiRepository.cs [hunk 15]
```diff
@@ -78,6 +92,13 @@ public async Task DeletePoiAsync(Guid id, CancellationToken token = default)
...
         }
 
+        if (!pois.Any())
+        {
+            _logger.LogInformation("This poi is not existing");
+        }
+        
+        _logger.LogInformation($"{pois.Count} matched poi(s) searched with Keyword: {keyword} (UserId : {userId})");
```
Comment (popemkt) [inline EzMap.Domain/Repositories/PoiRepository.cs:100]: Looking at this method, I can't see anywhere that it can return null, if it can't please adjust the return type to be suitable.

File: EzMap.Domain/Repositories/TagRepository.cs [hunk 16]
```diff
@@ -1,14 +1,15 @@
...
 
 namespace EzMap.Domain.Repositories;
 
 public interface ITagRepository
 {
     Task<Tag?> GetTagById(Guid? userId, Guid id, CancellationToken token = default);
 
-    void AddTag(TagCreateDto dto);
+    Guid AddTag(TagCreateDto dto);
     void UpdateTag(Tag? dbTag, TagUpdateDto dto);
```
Comment (popemkt) [inline EzMap.Domain/Repositories/TagRepository.cs]: Why does this method accept a nullable tag? in case it's null you just return? please make it so that methods should only accept values that it can operate on, not some value that cause it to short-circuit right away.

**Review by popemkt — CHANGES_REQUESTED** [review]

File: EzMap.Api/appsettings.Development.json [hunk 17]
```diff
@@ -2,7 +2,7 @@
   "Logging": {
     "LogLevel": {
       "Default": "Information",
-      "Microsoft.AspNetCore": "Warning"
```
Comment (popemkt) [inline EzMap.Api/appsettings.Development.json:5]: It should be Warning, or Info even

File: compose.yml [hunk 18]
```diff
@@ -44,8 +46,17 @@ services:
       - "5002:5601"
     depends_on:
       - es
+        
+  portainer:
```
Comment (popemkt) [inline compose.yml]: Not sure why docker compose for ezmap-poc needs portainer in the first place?

File: EzMap.Api/Controllers/PoiCollectionController.cs [hunk 19]
```diff
@@ -16,13 +19,46 @@ public class PoiCollectionController : ControllerBase
...
+                });
+            }
 
-        return await uow.SaveAsync() > 0
-            ? Ok("Your poi collection is created successfully!")
-            : new StatusCodeResult(StatusCodes.Status500InternalServerError);
+            await transaction.CommitAsync();
+            return Ok(new
+            {
+                Message = "Your poi collection is created and indexed successfully!",
```
Comment (popemkt) [inline EzMap.Api/Controllers/PoiCollectionController.cs]: Most 200 OK statuses don't need messages. The message shouldn't also leak implementation details (indexing), make it short, concise and minimal as possible. For error messages, try to also be concise, but add extra info on what's actually happening.

File: EzMap.Api/Controllers/PoiCollectionController.cs [hunk 20]
```diff
@@ -46,37 +82,97 @@ public async Task<IActionResult> Search([FromServices] IUnitOfWork uow,
...
+            {
+                return NotFound("Poi collection not found.");
+            }
+
+            uow.PoiCollectionRepository.UpdatePoiCollectionAsync(dbPoiCol, dto);
+            var dbResult = await uow.SaveAsync();
+            
+            if (dbResult == 0)
+            {
+                Console.WriteLine("No changes detected. Skipping commit.");
```
Comment (popemkt) [inline EzMap.Api/Controllers/PoiCollectionController.cs:100]: remove this

File: EzMap.Api/Controllers/PoiCollectionController.cs [hunk 21]
```diff
@@ -46,37 +82,97 @@ public async Task<IActionResult> Search([FromServices] IUnitOfWork uow,
...
+                return Ok("No changes were made to the POI collection.");
```
Comment (popemkt) [inline EzMap.Api/Controllers/PoiCollectionController.cs:101]: Ok doesn't need to have messages

File: EzMap.Api/Controllers/PoiCollectionController.cs (hunk 19 above)
Comment (popemkt) [reply EzMap.Api/Controllers/PoiCollectionController.cs]: It's also really inconsistent, this Ok returns a structured message: { Message, Name, Id }, while other Ok requests return only a plain string.

File: EzMap.Api/Controllers/PoiCollectionController.cs [hunk 22]
```diff
@@ -46,37 +82,97 @@ public async Task<IActionResult> Search([FromServices] IUnitOfWork uow,
...
+                await transaction.RollbackAsync();
+                return StatusCode(StatusCodes.Status500InternalServerError,
+                    "Failed to index the POI collection. The operation has been rolled back.");
+            }
 
-        return await uow.SaveAsync() > 0
-            ? Ok("Your poi collection is updated successfully")
-            : new StatusCodeResult(StatusCodes.Status500InternalServerError);
+            await transaction.CommitAsync();
+            return Ok("Your poi collection is updated successfully");
```
Comment (popemkt) [inline EzMap.Api/Controllers/PoiCollectionController.cs:113]: Same as above

File: EzMap.Api/Controllers/PoiCollectionController.cs [hunk 23]
```diff
@@ -46,37 +82,97 @@ public async Task<IActionResult> Search([FromServices] IUnitOfWork uow,
...
+                {
+                    Message =
+                        "Failed to remove POI collection from ElasticSearch. The operation has been rolled back.",
+                });
+            }
+
+            await transaction.CommitAsync();
+            return Ok(new
+            {
+                Message = "Your poi collection is deleted successfully!",
```
Comment (popemkt) [inline EzMap.Api/Controllers/PoiCollectionController.cs]: Same for inconsistency, this returns { Message }. Other places return plain string or other structure

File: EzMap.Api/Controllers/PoiController.cs [hunk 24]
```diff
@@ -17,47 +19,126 @@ public class PoiController : ControllerBase
...
+            var esResult = await elasticSearchService.AddOrUpdate(poiUpdateIndex);
+            if (!esResult)
+            {
+                await transaction.RollbackAsync();
+                return StatusCode(StatusCodes.Status500InternalServerError,
+                    "Failed to index the POI. The operation has been rolled back.");
+            }
+
+            await transaction.CommitAsync();
+            return Ok("Your poi is updated successfully!");
```
Comment (popemkt) [inline EzMap.Api/Controllers/PoiController.cs:88]: please fix every OK, I'm not gonna mark every time it happens with a comment

File: EzMap.Api/Program.cs [hunk 25]
```diff
@@ -12,8 +11,30 @@
...
+var environmentName = builder.Environment.EnvironmentName;
+
+var configuration = new ConfigurationBuilder()
+    .AddJsonFile("appsettings.json", optional: false, reloadOnChange: false)
+    .AddJsonFile($"appsettings.{environmentName}.json", optional: true, reloadOnChange: false)
+    .AddEnvironmentVariables()
+    .Build();
+
+
+
```
Comment (popemkt) [inline EzMap.Api/Program.cs]: don't create too many empty lines
